### PR TITLE
Update addons.rst

### PR DESCRIPTION
**Pull request by chrisseto**

### CODE REVIEW

**Comment by sloria** [comment]
The heading underlines are incorrect--the addon sections are showing up in the main table of contents

cos_development_ _cos_development_1_0_documentation

**Comment by chrisseto** [comment]
@sloria updated

**Comment by sloria** [comment]
Better, but still getting a bunch of warnings

1__sh

File: docs/osf/addons.rst [hunk 1]
```diff
@@ -31,31 +32,228 @@ During your installation you created a virtual environment for OSF. Switch to th
...
+StoredObject
+============
+
+A ``StoredObject`` is a class from our package ``modularodm`` that
+represents an object stored in a database.
+
+For our use case this database will always be TokuMX, a fork of mongodb
+that provides transactions and performance increases.
+
+Defining a ``StoredObject`` is very similar to the syntax used in
```
Comment (sloria) [inline docs/osf/addons.rst]: This is arguable. In general, this section might be too pedantic. I think it's enough to say:

```
StoredObject
==========

All models should be defined as subclasses of `framework.mongo.StoredObject`.

File: docs/osf/addons.rst [hunk 2]
```diff
@@ -31,31 +32,228 @@ During your installation you created a virtual environment for OSF. Switch to th
...
+`peewee`_ and `sqlalchemy`_
+
+The one caveat is that rather than using
+``modularodm.storedobject.StoredObject`` the base class must be
+``framework.mongo.StoredObject``
+
+Routes
+======
+
+A route is dictionary that containing one or more ``Rule`` objects in
```
Comment (sloria) [inline docs/osf/addons.rst]: > Routes are defined in a dictionary containing `rules` and an optional `prefix`.

File: docs/osf/addons.rst [hunk 3]
```diff
@@ -31,31 +32,228 @@ During your installation you created a virtual environment for OSF. Switch to th
...
+    @must_have_addon('myaddon', 'user')
+    def my_view(...):
+      pass
 
-**short_name (default is "s3")?**: Enter a lowercase, single word name. This name will be used in the file structure and as variables. You can use CamelCase if you need to have more words, or you can combine with numbers, but don't use any spaces or hyphens.
+The above code snippet will only run the view function if the specified
+model as the requested addon.
 
-**categories (default is "'storage'")?**: Enter a one word category, lowercase.
+Note: routes whose views are with decorated ``must_have_addon`` MUST start with ``/project/<pid>/...``.
```
Comment (sloria) [inline docs/osf/addons.rst]: Use 

```
.. note::
    Routes whose views...
```

File: docs/osf/addons.rst [hunk 4]
```diff
@@ -31,31 +32,228 @@ During your installation you created a virtual environment for OSF. Switch to th
...
 
+``from website.project.decorators.must_have_permission``
+--------------------------------------------------------
+
+``must_have_permission`` is another decorator factory, it takes a single
+permission argument (‘write’, ‘read’, ‘admin’).
```
Comment (sloria) [inline docs/osf/addons.rst]: Don't use smart quotes.

File: docs/osf/addons.rst [hunk 5]
```diff
@@ -31,31 +32,228 @@ During your installation you created a virtual environment for OSF. Switch to th
...
+The above code snippet will only run the view function if the specified
+model as the requested addon.
 
-**categories (default is "'storage'")?**: Enter a one word category, lowercase.
+Note: routes whose views are with decorated ``must_have_addon`` MUST start with ``/project/<pid>/...``.
 
+``from website.project.decorators.must_have_permission``
+--------------------------------------------------------
+
+``must_have_permission`` is another decorator factory, it takes a single
```
Comment (sloria) [inline docs/osf/addons.rst]: > ``must_have_permission` is another decorator factory that takes a `permission` argument (may be ``'write'``, ``'read'``, or ``'admin'``).

File: docs/osf/addons.rst [hunk 6]
```diff
@@ -78,13 +276,13 @@ Use the ``NodeLog`` class's named constants when possible,
 Every log action requires a template in ``youraddon/templates/log_templates.mako``. Each template's id corresponds to the name of the log action.
 
 
-Static files for addons
+Static files for add-ons
 ***********************
```
Comment (sloria) [inline docs/osf/addons.rst]: Make sure underline lengths are correct, else we get a bunch of warnings when building the docs.

Useful vim tip:

```
yypV*
```

File: docs/osf/addons.rst [hunk 7]
```diff
@@ -144,11 +340,11 @@ Builds the root or "dummy" folder for an addon.
 
         eg. Repo name for Github or bucket name for S3
 
-    :param permissions dict or Auth: Dictionary of permissions for the addon's content or Auth for use in node.can_X methods
+    :param permissions dict or Auth: Dictionary of permissions for the add-on's content or Auth for use in node.can_X methods
```
Comment (sloria) [inline docs/osf/addons.rst]: This was incorrect before, but could you fix this syntax? Should be

```
:param dict permissions: Dictionary of permissions... 
```

File: docs/osf/addons.rst [hunk 8]
```diff
@@ -144,11 +340,11 @@ Builds the root or "dummy" folder for an addon.
...
 
     :param urls dict: Hgrid related urls
 
-    :param extra String: Html to be appened to the addon folder name
+    :param extra String: Html to be appended to the addon folder name
```
Comment (sloria) [inline docs/osf/addons.rst]: Please fix this as well. Should be

```
:param str extra: HTML to be ...
```

File: docs/osf/addons.rst [hunk 9]
```diff
@@ -156,34 +352,16 @@ Builds the root or "dummy" folder for an addon.
...
-
-First requirement here is a render template, which should consist mainly of 1. the rendered file 2. a version history of the file
-
-Building File GUIDs
-*******************
-
-whenever a file is rendered a GUID should be created for it
-
-.. code-block:: python
+Some abstraction is in place to reduce redundancy for add-on that authorize access to third-party services via OAuth or OAuth2. Important classes to note include:
```
Comment (sloria) [inline docs/osf/addons.rst]: > There are utilities for add-ons that use OAuth or Oauth2 for authentication. These include:

File: docs/osf/addons.rst [hunk 10]
```diff
@@ -255,3 +433,13 @@ Example: When a Dropbox folder is shared on a project, contributors (and the pub
         ...
 
 Make sure that any view (CRUD, settings views...) that accesses resources from a 3rd-party service is secured in this way.
+
+
+.. _here: https://tools.ietf.org/html/rfc2119
```
Comment (sloria) [inline docs/osf/addons.rst]: Might want to inline this one:

```
... as defined `here <https://tools.ietf.org/html/rfc2119>`_.
```

File: docs/osf/addons.rst [hunk 11]
```diff
@@ -31,31 +32,228 @@ During your installation you created a virtual environment for OSF. Switch to th
...
+
+``from website.project.decorators.must_have_addon``
+---------------------------------------------------
+
+``must_have_addon`` is a decorator factory meaning you must supply
+arguments to it to get a decorator.
+
+.. code:: python
+
+    def must_have_addon(addon_short_name, model_name):
```
Comment (sloria) [inline docs/osf/addons.rst]: This line is unnecessary.

File: docs/osf/addons.rst [hunk 12]
```diff
@@ -6,16 +6,17 @@ Developing An Addon
 Notes and gotchas
 *****************
 
-- A new AddonSettings record is created upon enabling your addon's checkbox on the user settings page and submitting the form. You should *not* instantiate an `MyAddonUserSettings` object yourself
-- The view for rendering a file **must** include the return value of `website.project.util.serialize_project` (an alias of `_view_project`).
-- The view for rendering a file **must** use `website.addons.base.views.check_file_guid` and redirect if necessary
+- The words SHALL, MUST, MAY, etc are to be interpreted as defined `here`_
+- The add-on system is module based not class based
```
Comment (sloria) [inline docs/osf/addons.rst]: I'm not sure what is meant by this.

**Comment by sloria** [comment]
Pass finished.

File: docs/osf/addons.rst (hunk 11 above)
Comment (chrisseto) [inline docs/osf/addons.rst]: Its there just to sure the parameters it accepts.
Might be useful, might not?

Comment (sloria) [inline docs/osf/addons.rst]: It makes it look like devs are supposed to implement this function themselves. Better to show the import and 1 or 2 usage examples, as you have below.

**Comment by chrisseto** [comment]
Updated

File: docs/osf/addons.rst [hunk 13]
```diff
@@ -6,16 +6,17 @@ Developing An Addon
 Notes and gotchas
 *****************
 
-- A new AddonSettings record is created upon enabling your addon's checkbox on the user settings page and submitting the form. You should *not* instantiate an `MyAddonUserSettings` object yourself
-- The view for rendering a file **must** include the return value of `website.project.util.serialize_project` (an alias of `_view_project`).
-- The view for rendering a file **must** use `website.addons.base.views.check_file_guid` and redirect if necessary
+- The words SHALL, MUST, MAY, etc are to be interpreted as defined `here <https://tools.ietf.org/html/rfc2119>`_.
+- The add-on system is module based not class based
```
Comment (sloria) [inline docs/osf/addons.rst]: I'm still not sure what this means.

Comment (chrisseto) [inline docs/osf/addons.rst]: An addon is defined in/as a python module not a class

Comment (sloria) [inline docs/osf/addons.rst]: Hm, ok. This is probably unnecessary, but it 's fine.

File: docs/osf/addons.rst [hunk 14]
```diff
@@ -31,31 +32,218 @@ During your installation you created a virtual environment for OSF. Switch to th
...
+
+    @must_have_addon('myaddon', 'node')
+    def my_node_view(...):
+      pass
+
+The above code snippet will only run the view function if the specified
+model as the requested addon.
+
+.. note:: 
+    Routes whose views are with decorated ``must_have_addon`` MUST start with ``/project/<pid>/...``.
```
Comment (sloria) [inline docs/osf/addons.rst]: This is only true if the second argument is `node`. Wouldn't hurt to be explicit with `must_have_addon('addon_name', 'node')`.

File: docs/osf/addons.rst [hunk 15]
```diff
@@ -31,31 +32,218 @@ During your installation you created a virtual environment for OSF. Switch to th
...
+
+``from website.project.decorators.must_have_permission``
```
Comment (sloria) [inline docs/osf/addons.rst]: Remove "from"

File: docs/osf/addons.rst [hunk 16]
```diff
@@ -31,31 +32,218 @@ During your installation you created a virtual environment for OSF. Switch to th
...
 
+More can be found in ``website.project.decorators``.
+
+``framework.auth.decorators.must_be_logged_in``
+-----------------------------------------------
+
+Ensures that a user is logged in and imputes ``auth`` into keyword
+arguments
+
+``from website.project.decorators.must_have_addon``
```
Comment (sloria) [inline docs/osf/addons.rst]: Remove "from"

**Comment by sloria** [comment]
Pass finished.

**Comment by chrisseto** [comment]
Updated again

**Comment by sloria** [comment]
Formatting issue


developing_an_addon_ _cos_development_1_0_documentation

**Comment by sloria** [comment]
Will fix the above issue when I merge.

**Pull request merged**